### PR TITLE
JES API Manager: run creation and retries. Closes #1965. Closes #1914

### LINKED ISSUE #1914: Auto-retry transient HTTP based issues for Google interaction
When we interact with Google APIs, including JES and GCS sometimes things go awry for transient reasons. The belief is that retrying a bit would make it go green. We want to handle this automatically for the user. We shall:

- Retry up to N times via an exponential backoff. N is TBD (10?)
- Upon each retry, emit a log message
- If N is reached the job will fail w/ the last error event as the failure reason

TBD: Design discussion on how to implement this

### LINKED ISSUE #1965: FireCloud - cromwell losing jobs?
We are seeing infrequent situations where JES says a call is done but cromwell thinks it is still running.

I see the call starting in the logs:
```
2017-02-08 18:55:58,500 cromwell-system-akka.dispatchers.engine-dispatcher-963 INFO - WorkflowExecutionActor-fa7e25a2-f51f-4763-9f8a-5a2e5cd1c954 [UUID(fa7e25a2)]: Starting calls: pon_gatk_workflow.PadTargets:NA:1
```

Then the only suspicious things I see later in the logs are these messages (which could be completely unrelated however they do start to appear 2.5 minutes after the job completes on JES):
```
2017-02-08 19:17:07,588 cromwell-system-akka.dispatchers.backend-dispatcher-424 INFO - The JES polling actor Actor[akka://cromwell-system/user/cromwell-service/$b/$a/$Enn#762671444] unexpectedly terminated while conducting 100 polls. Making a new one...
java.lang.NullPointerException: null
2017-02-08 19:17:07,588 cromwell-system-akka.dispatchers.backend-dispatcher-246 ERROR - null
```

For future reference by a FireCloud admin the operations id is ELSl1PihKxjdhp-6gvr3weYBILma7PWMHyoPcHJvZHVjdGlvblF1ZXVl and the workflow id is fa7e25a2-f51f-4763-9f8a-5a2e5cd1c954 and the workflow was aborted 3PM Feb 8.

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/Run.scala [hunk 1]
```diff
@@ -81,120 +72,31 @@ object Run {
...
+    rpargs.setLogging(logging)
 
-  private def eventIfExists(name: String, metadata: Map[String, AnyRef], eventName: String): Option[ExecutionEvent] = {
-    metadata.get(name) map { time => ExecutionEvent(eventName, OffsetDateTime.parse(time.toString)) }
+    rpr
   }
 }
 
-case class Run(runId: String, genomicsInterface: Genomics) {
+final case class Run(runId: StandardAsyncJob, genomicsInterface: Genomics) {
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/Run.scala]: `runId` -> `job`?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala [hunk 2]
```diff
@@ -141,11 +107,9 @@ object JesPollingActor {
     * Forcing the minimum value to be 1 second, for now it seems unlikely to matter and it makes testing a bit
     * easier
     */
-  def determineBatchInterval(qps: Int): FiniteDuration = {
-    val maxInterval = MaxBatchSize / qps.toDouble // Force this to be floating point in case the value is < 1
+  def determineBatchInterval(qps: PositiveInt): FiniteDuration = {
+    val maxInterval = MaxBatchSize / qps.value.toDouble // Force this to be floating point in case the value is 0
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala]: Is this comment still relevant? `qps: PositiveInt` can never be zero. But perhaps one still wants `maxInterval` to be a double.

File: supportedBackends/jes/src/test/scala/cromwell/backend/impl/jes/statuspolling/RunSpec.scala [hunk 3]
```diff
@@ -19,27 +19,27 @@ class RunSpec extends FlatSpec with Matchers with MockitoTrait {
     val op: Operation = new Operation()
     op.setMetadata(eventsMetadata.asJava)
 
-    val list = Run.getEventList(op)
+    val list = StatusPolling.getEventList(op)
```
Comment (kshakir) [inline supportedBackends/jes/src/test/scala/cromwell/backend/impl/jes/statuspolling/RunSpec.scala]: Rename the spec and the behavior above to match the renamed main class:

```scala
class StatusPollingSpec extends FlatSpec with Matchers with MockitoTrait {
   behavior of "StatusPolling"
```

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala [hunk 4]
```diff
@@ -234,41 +241,58 @@ class JesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
...
 
-  protected def runWithJes(runIdForResumption: Option[String]): ExecutionHandle = {
-    // Force runtimeAttributes to evaluate so we can fail quickly now if we need to:
-    Try(runtimeAttributes) match {
-      case Success(_) =>
-        val jesInputs: Set[JesInput] = generateJesInputs(jobDescriptor) ++ monitoringScript + cmdInput
-        val jesOutputs: Set[JesFileOutput] = generateJesOutputs(jobDescriptor) ++ monitoringOutput
+  override def recoverAsync(jobId: StandardAsyncJob)(implicit ec: ExecutionContext): Future[ExecutionHandle] = {
+    runWithJes(Option(jobId))
+  }
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: Argh. Thanks!

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 5]
```diff
@@ -597,13 +597,13 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
       case exception: Exception =>
         val handler: PartialFunction[(ExecutionHandle, Exception), ExecutionHandle] =
           customPollStatusFailure orElse {
-            case (_: ExecutionHandle, exception: Exception) if isFatal(exception) =>
+            case (h: ExecutionHandle, exception: Exception) if isFatal(exception) =>
```
Comment (kshakir) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: `h` isn't used as far as I can tell?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala [hunk 6]
```diff
@@ -1,65 +1,84 @@
...
+    case failure: JesApiQueryFailed => handleQueryFailure(failure)
     case Terminated(actorRef) => handleTerminated(actorRef)
     case other => log.error(s"Unexpected message to JesPollingManager: $other")
   }
 
-  /**
-    * !! Function Impurity Warning: Modifies Actor Data !!
-    */
+  private def handleQueryFailure(failure: JesApiQueryFailed) = failure match {
+    case JesApiQueryFailed(request, cause) if request.failedAttempts < maxRetries =>
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: `JesApiQueryFailed(request, _)`

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala [hunk 7]
```diff
@@ -1,65 +1,84 @@
...
+      val nextRequest = request.withFailedAttempt
+      val delay = nextRequest.backoff.backoffMillis.millis
+      context.system.scheduler.scheduleOnce(delay, self, nextRequest)
+      ()
+    case JesApiQueryFailed(request, cause) =>
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: `JesApiQueryFailed(request, _)`

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesRunCreationClient.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+trait JesRunCreationClient { this: Actor with ActorLogging =>
+
+  private var runCreationClientPromise: Option[Promise[StandardAsyncJob]] = None
+
+  val pollingActor: ActorRef
+
+  def runCreationClientReceive: Actor.Receive = {
+    case runId: StandardAsyncJob =>
+      completePromise(Success(runId))
+    case JesApiQueryFailed(query, e) => completePromise(Failure(e))
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesRunCreationClient.scala]: `JesApiQueryFailed(_, e)`

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesStatusRequestClient.scala [hunk 9]
```diff
@@ -25,9 +23,9 @@ trait JesPollingActorClient { this: Actor with ActorLogging =>
     case r: RunStatus =>
       log.debug(s"Polled status received: $r")
       completePromise(Success(r))
-    case JesPollFailed(e, responseHeaders) =>
-      log.debug("JES poll failed! Sad.")
-      completePromise(Failure(new IOException(s"Google request failed: ${e.toPrettyString}")))
+    case JesApiQueryFailed(query, e) =>
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesStatusRequestClient.scala]: `JesApiQueryFailed(_, e)`

File: supportedBackends/jes/src/test/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActorSpec.scala [hunk 10]
```diff
@@ -148,9 +148,9 @@ class JesAsyncBackendJobExecutionActorSpec extends TestKitSuite("JesAsyncBackend
                              promise: Promise[BackendJobExecutionResponse],
                              jesSingletonActor: ActorRef): ActorRef = {
 
-    // Mock/stub out the bits that would reach out to JES.
-    val run = mock[Run]
-    val handle = new JesPendingExecutionHandle(jobDescriptor, StandardAsyncJob(run.runId), Option(run), None)
+    val runId = StandardAsyncJob(UUID.randomUUID().toString)
```
Comment (kshakir) [inline supportedBackends/jes/src/test/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActorSpec.scala]: `val job`?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala [hunk 11]
```diff
@@ -234,41 +241,58 @@ class JesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
...
+
+  private def runWithJes(runIdForResumption: Option[StandardAsyncJob]): Future[ExecutionHandle] = {
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: `runIdForResumption` => `jobForResumption`?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala [hunk 12]
```diff
@@ -234,41 +241,58 @@ class JesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
...
-        PendingExecutionHandle(jobDescriptor, StandardAsyncJob(run.runId), Option(run), previousStatus = None)
-      case Failure(e) => FailedNonRetryableExecutionHandle(e)
+      val jesParameters = standardParameters ++ gcsAuthParameter ++ jesInputs ++ jesOutputs
+
+      jobPaths.script.writeAsText(commandScriptContents)
+      createJesRunPipelineRequest(jesParameters)
+    }}
+
+    runIdForResumption match {
+      case Some(runId) =>
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: `Some(job)`?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesRunCreationClient.scala [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  *   override def receive = runCreationClientReceive orElse { ... }
+  */
+trait JesRunCreationClient { this: Actor with ActorLogging =>
+
+  private var runCreationClientPromise: Option[Promise[StandardAsyncJob]] = None
+
+  val pollingActor: ActorRef
+
+  def runCreationClientReceive: Actor.Receive = {
+    case runId: StandardAsyncJob =>
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesRunCreationClient.scala]: `runId` -> `job`?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesRunCreationClient.scala [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+      completePromise(Success(runId))
+    case JesApiQueryFailed(query, e) => completePromise(Failure(e))
+  }
+
+  private def completePromise(runId: Try[StandardAsyncJob]) = {
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesRunCreationClient.scala]: `runId` -> `job`?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/RunCreation.scala [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    val completionPromise = Promise[Try[Unit]]()
+    val resultHandler = runCreationResultHandler(runCreationQuery, completionPromise)
+    addRunCreationToBatch(runCreationQuery.rpr, runCreationQuery.genomicsInterface, batch, resultHandler)
+    completionPromise.future
+  }
+
+  private def addRunCreationToBatch(rpr: RunPipelineRequest, genomicsInterface: Genomics, batch: BatchRequest, resultHandler: JsonBatchCallback[Operation]) =
+    genomicsInterface.pipelines().run(rpr).queue(batch, resultHandler)
+
+  private def getRunId(operation: Operation) = StandardAsyncJob(operation.getName)
```
Comment (kshakir) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/RunCreation.scala]: `getJob`?

**Comment by kshakir** [comment]
Overall looks good, and if travis is happy I'm happy. 👍

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala [hunk 16]
```diff
@@ -90,4 +90,15 @@ object JesAttributes {
...
+  def validateQps(config: Config): ErrorOr[PositiveInt] = {
+    val qpsCandidate = config.as[Option[Int]]("genomics-api-queries-per-100-seconds").getOrElse(GenomicsApiDefaultQps) / 100
+
+    if (qpsCandidate > 0) PositiveInt(qpsCandidate).validNel
+    else "Supplied QPS for Google Genomics API was not positive, value was {} using {} instead".invalidNel
+  }
+}
+
+final case class PositiveInt(value: Int) {
+  if (value < 0) throw new RuntimeException("PositiveInt cannot have a value below 0!")
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala]: We had a convo already about my thoughts on PositiveInt. Either use something real or ditch the idea in general.

**Comment by geoffjentry** [comment]
Have you run this under heavy load? If so, how does it react?

File: core/src/main/scala/cromwell/util/StopAndLogSupervisor.scala [hunk 17]
```diff
@@ -2,9 +2,8 @@ package cromwell.util
 
 import akka.actor.SupervisorStrategy.{Decider, Stop}
 import akka.actor.{Actor, ActorRef, OneForOneStrategy, SupervisorStrategy}
-import cromwell.core.logging.WorkflowLogging
 
-trait StopAndLogSupervisor { this: Actor with WorkflowLogging =>
+trait StopAndLogSupervisor { this: Actor =>
```
Comment (geoffjentry) [inline core/src/main/scala/cromwell/util/StopAndLogSupervisor.scala]: TIL this was a thing. Thanks for using it so I could be educated :)

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala [hunk 18]
```diff
@@ -71,12 +79,12 @@ class JesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
...
 
   override def tryAbort(job: StandardAsyncJob): Unit = {
-    Run(job.jobId, initializationData.genomics).abort()
+    Run(job, initializationData.genomics).abort()
   }
 
   override def requestsAbortAndDiesImmediately: Boolean = true
 
-  override def receive: Receive = pollingActorClientReceive orElse super.receive
+  override def receive: Receive = pollingActorClientReceive orElse runCreationClientReceive orElse super.receive
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala]: tol as they say. I know they're all over the place now but i find these layered receives to be really hard to follow when i'm tracking down message flow

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala (hunk 16 above)
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala]: Oh yeah, I just forgot by the time I started looking at this again 😳

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesConfiguration.scala [hunk 19]
```diff
@@ -33,6 +33,6 @@ class JesConfiguration(val configurationDescriptor: BackendConfigurationDescript
   val genomicsFactory = GenomicsFactory(googleConfig.applicationName, jesAuths.genomics, jesAttributes.endpointUrl)
   val dockerCredentials = DockerConfiguration.build(configurationDescriptor.backendConfig).dockerCredentials map JesDockerCredentials.apply
   val needAuthFileUpload = jesAuths.gcs.requiresAuthFile || dockerCredentials.isDefined
-  val qps = jesAttributes.qps
+  val qps: PositiveInt = jesAttributes.qps
```
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesConfiguration.scala]: I may or may not sometimes use ascriptions and the subsequent build failures to detect usage...

File: project/Dependencies.scala [hunk 20]
```diff
@@ -111,6 +111,8 @@ object Dependencies {
   val databaseSqlDependencies = baseDependencies ++ slickDependencies ++ dbmsDependencies
 
   val coreDependencies = List(
+    "org.scala-lang" % "scala-compiler" % Settings.ScalaVersion % Provided,
+    "eu.timepit" %% "refined"            % "0.7.0",
```
Comment (geoffjentry) [inline project/Dependencies.scala]: what's up w/ the spacing here. if you're trying to line stuff up (and nothing else is), it's not lined up right.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala [hunk 21]
```diff
@@ -24,7 +26,7 @@ case class JesAttributes(project: String,
                          executionBucket: String,
                          endpointUrl: URL,
                          maxPollingInterval: Int,
-                         qps: Int,
+                         qps: Int Refined Positive,
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala]: oh-yeah-t-shirt-doomsday-attire2-250x316

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala [hunk 22]
```diff
@@ -3,19 +3,21 @@ package cromwell.backend.impl.jes
 import akka.actor.{Actor, ActorLogging, Props}
 import cromwell.core.Dispatcher.BackendDispatcher
 import cromwell.backend.impl.jes.statuspolling.JesApiQueryManager
-import cromwell.backend.impl.jes.statuspolling.JesApiQueryManager.DoPoll
+import cromwell.backend.impl.jes.statuspolling.JesApiQueryManager.JesApiQueryManagerRequest
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.numeric.Positive
 
-final case class JesBackendSingletonActor(qps: Int) extends Actor with ActorLogging {
+final case class JesBackendSingletonActor(qps: Int Refined Positive) extends Actor with ActorLogging {
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala]: I wouldn't be opposed to a typedef `PositiveInt` to `Int Refined Positive` in a relevant package object but I also am 100% AOK w/ it staying like this. 

Phrased this way as I often am against such things ;)

File: project/Dependencies.scala (hunk 20 above)
Comment (cjllanwarne) [inline project/Dependencies.scala]: Bad copy/paste I think? Not something I intended!

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala (hunk 22 above)
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala]: or in a companion object if it's contained solely in this file (he says w/o looking)

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala [hunk 23]
```diff
@@ -72,44 +93,77 @@ class JesApiQueryManager(val qps: Int) extends Actor with ActorLogging {
   }
 
   private def handleTerminated(terminee: ActorRef) = {
-    // Currently we can assume this is a polling actor. Might change in a future update:
+    val cause = getFailureCause(terminee).getOrElse(new RuntimeException("No failure reason recorded"))
+    // We assume this is a polling actor. Might change in a future update:
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: How are we protected from pathological cases like the NPE we saw?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala [hunk 24]
```diff
@@ -3,88 +3,65 @@ package cromwell.backend.impl.jes.statuspolling
...
       ()
     case NoWorkToDo =>
       scheduleCheckForWork()
   }
 
-  private def handleBatch(workBatch: NonEmptyList[JesStatusPollQuery]): NonEmptyList[Future[Try[Unit]]] = {
-
-    // Assume that the auth for the first element is also able to query the remaining Runs
-    val batch: BatchRequest = createBatch(workBatch.head.run)
+  private def handleBatch(workBatch: NonEmptyList[JesApiQuery]): Future[List[Try[Unit]]] = {
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala]: Presumably less feasible now due to the inversion of the list & future, but since futures already contain a try could this be simplified?

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala [hunk 25]
```diff
@@ -93,14 +70,15 @@ class JesPollingActor(val pollingManager: ActorRef, val qps: Int) extends Actor
...
+        log.warning("{} failures (from {} requests) fetching JES statuses", errors.size, someFailures.size)
       } else {
-        log.warning("{} failures fetching JES statuses: {}", errors.size, errors.mkString(", "))
+        log.warning("{} failures (from {} requests) fetching JES statuses: {}", errors.size, someFailures.size, errors.mkString(", "))
       }
       scheduleCheckForWork()
     case Failure(t) =>
       // NB: Should be impossible since we only ever do completionPromise.trySuccess()
-      log.error("Completion promise unexpectedly set to Failure: {}", t.getMessage)
+      val msg = "Programmer Error: Completion promise unexpectedly set to Failure: {}. Don't do this, otherwise the Future.sequence is short-circuited on the first failure"
```
Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala]: i laughed

**Comment by geoffjentry** [comment]
@cjllanwarne You asked at standup what I was asking for when I asked about how it handles load. What I meant was if you submit stuff such that there are at least in the range of several thousand runs & status polls happening (ideally interspersed as well, not just a bunch of runs followed by a bunch of statuses) does something terrible happen? 

IOW a week from now when thousands of things are flowing through via FC is there some obvious gotcha that we could have spotted now to save us some pain.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala (hunk 22 above)
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala]: Hmm, I think I prefer leaving it raw but only because
 - I tend to dislike typedefs too
 - Putting on my evangelizing hat, people are more likely to see that it's an awesome general purpose library in this raw form.

I'll leave this up to you... (with silence meaning "probably won't change")

Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesBackendSingletonActor.scala]: Oh I was just commenting that my normal objections don't apply here so if by any chance you had originally said "I want to make a `type` but mean @geoffjentry would hate it!" to feel free. Although who am I kidding, that never stops anyone ;)

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala (hunk 23 above)
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: - The Polling actor dies
 - The supervisor puts it back into the queue with an incremented failure count

Same as any other failure, except it's resubmitted due to `Terminated(ref)`, rather than `JesApiFailure(...)`

Comment (geoffjentry) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: Ah right, I forgot about the reason for this in the first place, the cap on retries

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala (hunk 25 above)
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala]: ... unless we see "Programmer Error" showing up in the production logs...! 😨

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala [hunk 26]
```diff
@@ -90,4 +92,16 @@ object JesAttributes {
...
   }
+
+  def validateQps(config: Config): ErrorOr[Int Refined Positive] = {
+    import eu.timepit.refined._
+
+    val qp100s = config.as[Option[Int]]("genomics-api-queries-per-100-seconds").getOrElse(GenomicsApiDefaultQps)
+    val qpsCandidate = qp100s / 100
+
+    refineV[Positive](qpsCandidate) match {
+      case Left(_) => s"Supplied QPS for Google Genomics API was not a positive integer (supplied value was $qpsCandidate".invalidNel
```
Comment (Horneth) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAttributes.scala]: `)` missing at the end. Also the actual supplied value was `qp100s` not `qpsCandidate` no ? It might be confusing if someone get this error message, look at their conf and see a different number.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala [hunk 27]
```diff
@@ -1,65 +1,86 @@
...
+    case failure: JesApiQueryFailed => handleQueryFailure(failure)
     case Terminated(actorRef) => handleTerminated(actorRef)
     case other => log.error(s"Unexpected message to JesPollingManager: $other")
   }
 
-  /**
-    * !! Function Impurity Warning: Modifies Actor Data !!
-    */
+  private def handleQueryFailure(failure: JesApiQueryFailed) = failure match {
+    case JesApiQueryFailed(request, _) if request.failedAttempts < maxRetries =>
```
Comment (Horneth) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: TOL: all failures, as long as they are under the `maxRetries` count, are considered retryable ? 
To be clear I don't think it's a big deal because even if someone messes up their conf and get `Forbidden` responses nothing will start and they will have to stop cromwell anyway but just as a note.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala [hunk 28]
```diff
@@ -3,88 +3,65 @@ package cromwell.backend.impl.jes.statuspolling
...
       scheduleCheckForWork()
   }
 
-  private def handleBatch(workBatch: NonEmptyList[JesStatusPollQuery]): NonEmptyList[Future[Try[Unit]]] = {
-
-    // Assume that the auth for the first element is also able to query the remaining Runs
-    val batch: BatchRequest = createBatch(workBatch.head.run)
+  private def handleBatch(workBatch: NonEmptyList[JesApiQuery]): Future[List[Try[Unit]]] = {
+    // Assume that the auth for the first element is also good enough for everything else:
+    val batch: BatchRequest = createBatch(workBatch.head.genomicsInterface)
```
Comment (Horneth) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesPollingActor.scala]: TOL - for later: I think the proper way to do this would be to create a batch request from scratch, with no auth.
The batch request itself should not *need* to be authenticated, and therefore should not be, as all the sub requests should have their own Http authentication header. However I just tested it and it seems that the google library fails to parse the response if there's not auth in the batch request.
Also from the doc, it looks like we can batch 1000 requests per batch for the genomics api: https://cloud.google.com/genomics/batch

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/RunCreation.scala [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    val resultHandler = runCreationResultHandler(runCreationQuery, completionPromise)
+    addRunCreationToBatch(runCreationQuery.rpr, runCreationQuery.genomicsInterface, batch, resultHandler)
+    completionPromise.future
+  }
+
+  private def addRunCreationToBatch(rpr: RunPipelineRequest, genomicsInterface: Genomics, batch: BatchRequest, resultHandler: JsonBatchCallback[Operation]) =
+    genomicsInterface.pipelines().run(rpr).queue(batch, resultHandler)
+
+  private def getJob(operation: Operation) = StandardAsyncJob(operation.getName)
+}
```
Comment (Horneth) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/RunCreation.scala]: new line

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/StatusPolling.scala [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+        case (None, None) =>
+          throw new IllegalArgumentException("Both jesReportedEndTime and finalEventsListTime were None.")
+      }
+    }
+  }
+
+  private def eventIfExists(name: String, metadata: Map[String, AnyRef], eventName: String): Option[ExecutionEvent] = {
+    metadata.get(name) map { time => ExecutionEvent(eventName, OffsetDateTime.parse(time.toString)) }
+  }
+}
```
Comment (Horneth) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/StatusPolling.scala]: new line

**Comment by Horneth** [comment]
👍

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala (hunk 27 above)
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/JesApiQueryManager.scala]: The NPEs are particularly nasty as (previously) they'd bring down an entire batch (i.e. 100, of which only 1 was bad). With @kshakir's changes, only one should now fail and it won't go through the actor death mechanism, but the catch-all mechanism still shouldn't blanket destroy all 100 batched messages just because 1 was bad.

File: supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/RunCreation.scala (hunk 29 above)
Comment (cjllanwarne) [inline supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/statuspolling/RunCreation.scala]: Done
